tests/js/lib/link-checker.js

```diff
@@ -0,0 +1,166 @@
+/*
+
+    A component that trawls our generated content and reports errors if there are:
+
+    1. Broken links between pages on the site.
+    2. Broken links within a single page.
+
+ */
+/* eslint-env node */
+"use strict";
+var fluid   = require("infusion");
+var request = require("request");
+var url     = require("url");
+var jsdom   = require("jsdom");
+
+fluid.registerNamespace("fluid.tests.docs.linkChecker");
+
+fluid.tests.docs.linkChecker.startScan = function (that) {
+    var startingUrl = fluid.tests.docs.linkChecker.resolveSafely(that.options.baseUrl, that.options.startPage);
+
+    that.pagesToScan.push(startingUrl);
+
+    that.scanSinglePageOrFinish();
+};
+
+fluid.tests.docs.linkChecker.scanSinglePageOrFinish = function (that) {
+    if (that.pagesToScan.length) {
+        var promise = fluid.promise();
+        promise.then(that.scanSinglePageOrFinish);
+
+        var pageToScan = that.pagesToScan.pop();
+        if (that.scannedPages[pageToScan]) {
+            // Skip this, we've already scanned it.
+            promise.resolve();
+        }
+        else {
+            // Retrieve the page and process it.
+            request(pageToScan, function (error, response, body) {
+                fluid.tests.docs.linkChecker.processSingleResponse(that, pageToScan, error, response, body);
+                promise.resolve();
+            });
+        }
+    }
+    else {
+        fluid.tests.docs.linkChecker.checkInternalLinks(that);
+        fluid.tests.docs.linkChecker.checkPageLinks(that);
+
+        var report = fluid.model.transformWithRules(that, that.options.rules.report);
+        that.events.onResultsAvailable.fire(report);
+    }
+};
+
+fluid.tests.docs.linkChecker.urlBeforeHash = function (link) {
+    return link.indexOf("#") !== -1 ? link.substring(0, link.indexOf("#")) : link;
+};
+
+fluid.tests.docs.linkChecker.urlAfterHash = function (link) {
+    return link.indexOf("#") !== -1 ? link.substring(link.indexOf("#") + 1) : "";
+};
+
+fluid.tests.docs.linkChecker.checkInternalLinks = function (that) {
+    fluid.each(that.internalLinksToCheck, function (link) {
+        var linkBeforeHash = fluid.tests.docs.linkChecker.urlBeforeHash(link);
+        var linkAfterHash  = fluid.tests.docs.linkChecker.urlAfterHash(link);
+        if (linkAfterHash && linkAfterHash.length) {
+            var pageIds = that.idsByPage[linkBeforeHash];
+            if (!pageIds || pageIds.indexOf(linkAfterHash) === -1) {
+                that.brokenInternalLinks.push(link);
```

Comment:
Could we add a little extra tracking in this outgoing structure (which I guess will need to be pushed through some arguments) so that the user at least has the courtesy of knowing which file the broken link appears in? I appreciate that the location within the file will be less useful since this is generated content

Reply:
I handled this a different way but agree with and addressed the goal, which is to assist in actually fixing the broken link.